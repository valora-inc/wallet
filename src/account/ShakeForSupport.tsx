import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import Button, { BtnTypes } from 'src/components/Button'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'

export default function ShakeForSupport() {
  const { t } = useTranslation()

  const onContactSupport = () => {
    navigate(Screens.SupportContact)
  }

  return (
    <SafeAreaView style={styles.shakeForSupport}>
      <Text testID="HavingTrouble" style={styles.supportTitle}>
        {t('havingTrouble')}
      </Text>
      <Text testID="ShakeForSupport" style={styles.supportSubtitle}>
        {t('shakeForSupport')}
      </Text>
      <Button
        onPress={onContactSupport}
        text={t('contactSupport')}
        type={BtnTypes.PRIMARY}
        testID="ContactSupportFromShake"
      />
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  shakeForSupport: {
    position: 'absolute',
    bottom: 0,
    width: '100%',
    backgroundColor: colors.light,
    borderTopLeftRadius: 16,
    borderTopRightRadius: 16,
    padding: 24,
    alignItems: 'center',
  },
  supportTitle: {
    ...fontStyles.h2,
    marginTop: 16,
  },
  supportSubtitle: {
    ...fontStyles.regular,
    marginTop: 8,
    marginBottom: 16,
    textAlign: 'center',
  },
})
