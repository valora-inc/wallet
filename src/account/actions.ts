import { PincodeType } from 'src/account/reducer'

export enum Actions {
  CHOOSE_CREATE_ACCOUNT = 'ACCOUNT/CHOOSE_CREATE',
  CHOOSE_RESTORE_ACCOUNT = 'ACCOUNT/CHOOSE_RESTORE',
  START_STORE_WIPE_RECOVERY = 'ACCOUNT/START_STORE_WIPE_RECOVERY',
  CANCEL_CREATE_OR_RESTORE_ACCOUNT = 'ACCOUNT/CANCEL_CREATE_OR_RESTORE_ACCOUNT',
  SET_NAME = 'ACCOUNT/SET_NAME',
  SET_PHONE_NUMBER = 'ACCOUNT/SET_PHONE_NUMBER',
  SAVE_NAME = 'ACCOUNT/SAVE_NAME',
  DEV_MODE_TRIGGER_CLICKED = 'ACCOUNT/NAME_CLICKED',
  PHOTOSNUX_CLICKED = 'ACCOUNT/PHOTOSNUX_CLICKED',
  SET_PINCODE_SUCCESS = 'ACCOUNT/SET_PINCODE_SUCCESS',
  SET_PINCODE_FAILURE = 'ACCOUNT/SET_PINCODE_FAILURE',
  SET_ACCOUNT_CREATION_TIME = 'ACCOUNT/SET_ACCOUNT_CREATION_TIME',
  INITIALIZE_ACCOUNT = 'ACCOUNT/INITIALIZE_ACCOUNT',
  INITIALIZE_ACCOUNT_SUCCESS = 'ACCOUNT/INITIALIZE_ACCOUNT_SUCCESS',
  INITIALIZE_ACCOUNT_FAILURE = 'ACCOUNT/INITIALIZE_ACCOUNT_FAILURE',
  SET_BACKUP_COMPLETED = 'ACCOUNT/SET_BACKUP_COMPLETED',
  DISMISS_GET_VERIFIED = 'ACCOUNT/DISMISS_GET_VERIFIED',
  DISMISS_GOLD_EDUCATION = 'ACCOUNT/DISMISS_GOLD_EDUCATION',
  SET_USER_CONTACT_DETAILS = 'ACCOUNT/SET_USER_CONTACT_DETAILS',
  ACCEPT_TERMS = 'ACCOUNT/ACCEPT_TERMS',
  CLEAR_STORED_ACCOUNT = 'ACCOUNT/CLEAR_STORED_ACCOUNT',
  PROFILE_UPLOADED = 'ACCOUNT/PROFILE_UPLOADED',
  SAVE_SIGNED_MESSAGE = 'ACCOUNT/SAVE_SIGNED_MESSAGE',
  SET_CELO_EDUCATION_COMPLETED = 'ACCOUNT/SET_CELO_EDUCATION_COMPLETED',
  RECOVERY_PHRASE_IN_ONBOARDING_STARTED = 'ACCOUNT/RECOVERY_PHRASE_IN_ONBOARDING_STARTED',
  RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED = 'ACCOUNT/RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED',
}

export interface ChooseCreateAccountAction {
  type: Actions.CHOOSE_CREATE_ACCOUNT
  now: number
}
interface ChooseRestoreAccountAction {
  type: Actions.CHOOSE_RESTORE_ACCOUNT
}

interface StartStoreWipeRecoveryAction {
  type: Actions.START_STORE_WIPE_RECOVERY
  accountToRecover: string
}

interface CancelCreateOrRestoreAccountAction {
  type: Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT
}

interface SetNameAction {
  type: Actions.SET_NAME
  name: string
}

interface SetPhoneNumberAction {
  type: Actions.SET_PHONE_NUMBER
  e164PhoneNumber: string
  countryCode: string
}

interface SaveNameAction {
  type: Actions.SAVE_NAME
  name: string
}
interface DevModeTriggerClickedAction {
  type: Actions.DEV_MODE_TRIGGER_CLICKED
}

interface AcceptTermsAction {
  type: Actions.ACCEPT_TERMS
}

interface PhotosNUXClickedAction {
  type: Actions.PHOTOSNUX_CLICKED
}

interface SetPincodeSuccessAction {
  type: Actions.SET_PINCODE_SUCCESS
  pincodeType: PincodeType
}

interface SetPincodeFailureAction {
  type: Actions.SET_PINCODE_FAILURE
}

interface InitializeAccountAction {
  type: Actions.INITIALIZE_ACCOUNT
}

interface InitializeAccountSuccessAction {
  type: Actions.INITIALIZE_ACCOUNT_SUCCESS
}

interface SetAccountCreationAction {
  type: Actions.SET_ACCOUNT_CREATION_TIME
  now: number
}

interface SetBackupCompletedAction {
  type: Actions.SET_BACKUP_COMPLETED
}

interface DismissGetVerifiedAction {
  type: Actions.DISMISS_GET_VERIFIED
}

interface DismissGoldEducationAction {
  type: Actions.DISMISS_GOLD_EDUCATION
}

interface SetContactDetailsAction {
  type: Actions.SET_USER_CONTACT_DETAILS
  contactId: string
  thumbnailPath: string | null
}

export interface ClearStoredAccountAction {
  type: Actions.CLEAR_STORED_ACCOUNT
  account: string
}

interface ProfileUploadedAction {
  type: Actions.PROFILE_UPLOADED
}

interface SaveSignedMessage {
  type: Actions.SAVE_SIGNED_MESSAGE
}

interface SetCeloEducationCompletedAction {
  type: Actions.SET_CELO_EDUCATION_COMPLETED
  celoEducationCompleted: boolean
}

interface RecoveryPhraseInOnboardingStarted {
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED
}

interface RecoveryPhraseInOnboardingCompleted {
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED
}

export type ActionTypes =
  | ChooseCreateAccountAction
  | ChooseRestoreAccountAction
  | StartStoreWipeRecoveryAction
  | CancelCreateOrRestoreAccountAction
  | SetNameAction
  | SetPhoneNumberAction
  | SaveNameAction
  | DevModeTriggerClickedAction
  | PhotosNUXClickedAction
  | SetPincodeSuccessAction
  | SetPincodeFailureAction
  | InitializeAccountAction
  | InitializeAccountSuccessAction
  | SetAccountCreationAction
  | SetBackupCompletedAction
  | DismissGetVerifiedAction
  | DismissGoldEducationAction
  | SetContactDetailsAction
  | AcceptTermsAction
  | ClearStoredAccountAction
  | ProfileUploadedAction
  | SaveSignedMessage
  | SetCeloEducationCompletedAction
  | RecoveryPhraseInOnboardingStarted
  | RecoveryPhraseInOnboardingCompleted

export function chooseCreateAccount(now: number): ChooseCreateAccountAction {
  return {
    type: Actions.CHOOSE_CREATE_ACCOUNT,
    now,
  }
}

export function chooseRestoreAccount(): ChooseRestoreAccountAction {
  return {
    type: Actions.CHOOSE_RESTORE_ACCOUNT,
  }
}

export function startStoreWipeRecovery(accountToRecover: string): StartStoreWipeRecoveryAction {
  return {
    type: Actions.START_STORE_WIPE_RECOVERY,
    accountToRecover,
  }
}

export function cancelCreateOrRestoreAccount(): CancelCreateOrRestoreAccountAction {
  return {
    type: Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT,
  }
}

export function acceptTerms(): AcceptTermsAction {
  return {
    type: Actions.ACCEPT_TERMS,
  }
}

export function saveName(name: string): SaveNameAction {
  return {
    type: Actions.SAVE_NAME,
    name,
  }
}
export const devModeTriggerClicked = (): DevModeTriggerClickedAction => ({
  type: Actions.DEV_MODE_TRIGGER_CLICKED,
})

export const setPincodeSuccess = (pincodeType: PincodeType): SetPincodeSuccessAction => ({
  type: Actions.SET_PINCODE_SUCCESS,
  pincodeType,
})

export const initializeAccount = (): InitializeAccountAction => ({
  type: Actions.INITIALIZE_ACCOUNT,
})

export const initializeAccountSuccess = (): InitializeAccountSuccessAction => ({
  type: Actions.INITIALIZE_ACCOUNT_SUCCESS,
})

export const setAccountCreationTime = (now: number): SetAccountCreationAction => ({
  type: Actions.SET_ACCOUNT_CREATION_TIME,
  now,
})

export const setBackupCompleted = (): SetBackupCompletedAction => ({
  type: Actions.SET_BACKUP_COMPLETED,
})

export const dismissGetVerified = (): DismissGetVerifiedAction => ({
  type: Actions.DISMISS_GET_VERIFIED,
})

export const dismissGoldEducation = (): DismissGoldEducationAction => ({
  type: Actions.DISMISS_GOLD_EDUCATION,
})

export const setUserContactDetails = (
  contactId: string,
  thumbnailPath: string | null
): SetContactDetailsAction => ({
  type: Actions.SET_USER_CONTACT_DETAILS,
  contactId,
  thumbnailPath,
})

export const clearStoredAccount = (account: string): ClearStoredAccountAction => ({
  type: Actions.CLEAR_STORED_ACCOUNT,
  account,
})

export const saveSignedMessage = (): SaveSignedMessage => ({
  type: Actions.SAVE_SIGNED_MESSAGE,
})

export const setGoldEducationCompleted = (): SetCeloEducationCompletedAction => ({
  type: Actions.SET_CELO_EDUCATION_COMPLETED,
  celoEducationCompleted: true,
})

export const recoveryPhraseInOnboardingStarted = (): RecoveryPhraseInOnboardingStarted => ({
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED,
})

export const recoveryPhraseInOnboardingCompleted = (): RecoveryPhraseInOnboardingCompleted => ({
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED,
})
