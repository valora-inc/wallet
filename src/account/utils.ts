import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
