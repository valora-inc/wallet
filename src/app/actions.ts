import { BIOMETRY_TYPE } from 'react-native-keychain'
import { RemoteConfigValues } from 'src/app/saga'
import { Screens } from 'src/navigator/Screens'

// https://facebook.github.io/react-native/docs/appstate
export enum AppState {
  Background = 'Background',
  Active = 'Active',
  Inactive = 'Inactive',
}

export enum MultichainBetaStatus {
  NotSeen = 'NotSeen',
  OptedIn = 'OptedIn',
  OptedOut = 'OptedOut',
}

export enum Actions {
  SET_APP_STATE = 'APP/SET_APP_STATE',
  SET_LOGGED_IN = 'APP/SET_LOGGED_IN',
  SET_SUPPORTED_BIOMETRY_TYPE = 'APP/SET_SUPPORTED_BIOMETRY_TYPE',
  OPEN_DEEP_LINK = 'APP/OPEN_DEEP_LINK',
  DEEP_LINK_DEFERRED = 'APP/DEEP_LINK_DEFERRED',
  SET_ANALYTICS_ENABLED = 'APP/SET_ANALYTICS_ENABLED',
  SET_LOCK_WITH_PIN_ENABLED = 'APP/SET_LOCK_WITH_PIN_ENABLED',
  LOCK = 'APP/LOCK',
  UNLOCK = 'APP/UNLOCK',
  SET_SESSION_ID = 'SET_SESSION_ID',
  OPEN_URL = 'APP/OPEN_URL',
  UPDATE_REMOTE_CONFIG_VALUES = 'APP/UPDATE_REMOTE_CONFIG_VALUES',
  ACTIVE_SCREEN_CHANGED = 'APP/ACTIVE_SCREEN_CHANGED',
  APP_MOUNTED = 'APP/APP_MOUNTED',
  APP_UNMOUNTED = 'APP/APP_UNMOUNTED',
  ANDROID_MOBILE_SERVICES_AVAILABILITY_CHECKED = 'APP/ANDROID_MOBILE_SERVICES_AVAILABILITY_CHECKED',
  PHONE_NUMBER_VERIFICATION_COMPLETED = 'APP/PHONE_NUMBER_VERIFICATION_COMPLETED',
  PHONE_NUMBER_REVOKED = 'APP/PHONE_NUMBER_REVOKED',
  INVITE_LINK_CONSUMED = 'APP/INVITE_LINK_CONSUMED',
  HAPTIC_FEEDBACK_SET = 'APP/HAPTIC_FEEDBACK_SET',
  PUSH_NOTIFICATIONS_PERMISSION_CHANGED = 'APP/PUSH_NOTIFICATIONS_PERMISSION_CHANGED',
  IN_APP_REVIEW_REQUESTED = 'APP/IN_APP_REVIEW_REQUESTED',
  NOTIFICATION_SPOTLIGHT_SEEN = 'APP/NOTIFICATION_SPOTLIGHT_SEEN',
  TOGGLE_HIDE_BALANCES = 'APP/TOGGLE_HIDE_BALANCES',
  OPT_MULTICHAIN_BETA = 'APP/OPT_MULTICHAIN_BETA',
}

export interface SetAppState {
  type: Actions.SET_APP_STATE
  state: string
}

interface SetLoggedIn {
  type: Actions.SET_LOGGED_IN
  loggedIn: boolean
}

interface SetSupportedBiometryType {
  type: Actions.SET_SUPPORTED_BIOMETRY_TYPE
  supportedBiometryType: BIOMETRY_TYPE | null
}

export interface OpenDeepLink {
  type: Actions.OPEN_DEEP_LINK
  deepLink: string
  isSecureOrigin: boolean
}

interface DeepLinkDeferred {
  type: Actions.DEEP_LINK_DEFERRED
  deepLink: string
  isSecureOrigin: boolean
}

interface SetAnalyticsEnabled {
  type: Actions.SET_ANALYTICS_ENABLED
  enabled: boolean
}

interface SetRequirePinOnAppOpen {
  type: Actions.SET_LOCK_WITH_PIN_ENABLED
  enabled: boolean
}

interface ActiveScreenChangedAction {
  type: Actions.ACTIVE_SCREEN_CHANGED
  activeScreen: Screens
}

export interface Lock {
  type: Actions.LOCK
}

export interface Unlock {
  type: Actions.UNLOCK
}

export interface AppMounted {
  type: Actions.APP_MOUNTED
}

export interface AppUnmounted {
  type: Actions.APP_UNMOUNTED
}

export interface SetSessionId {
  type: Actions.SET_SESSION_ID
  sessionId: string
}

export interface OpenUrlAction {
  type: Actions.OPEN_URL
  url: string
  openExternal: boolean
  isSecureOrigin: boolean
}

export interface UpdateConfigValuesAction {
  type: Actions.UPDATE_REMOTE_CONFIG_VALUES
  configValues: RemoteConfigValues
}

export interface AndroidMobileServicesAvailabilityChecked {
  type: Actions.ANDROID_MOBILE_SERVICES_AVAILABILITY_CHECKED
  googleIsAvailable: boolean | undefined
  huaweiIsAvailable: boolean | undefined
}

export interface PhoneNumberVerificationCompleted {
  type: Actions.PHONE_NUMBER_VERIFICATION_COMPLETED
  e164PhoneNumber: string
  countryCode: string | null
}

export interface PhoneNumberRevoked {
  type: Actions.PHONE_NUMBER_REVOKED
  e164PhoneNumber: string
}

export interface InviteLinkConsumed {
  type: Actions.INVITE_LINK_CONSUMED
  inviterAddress: string
}

interface HapticFeedbackSet {
  type: Actions.HAPTIC_FEEDBACK_SET
  hapticFeedbackEnabled: boolean
}

export interface PushNotificationsPermissionChanged {
  type: Actions.PUSH_NOTIFICATIONS_PERMISSION_CHANGED
  enabled: boolean
  requestedInApp: boolean
}

export interface inAppReviewRequested {
  type: Actions.IN_APP_REVIEW_REQUESTED
  inAppReviewLastInteractionTimestamp: number | null
}

export interface NotificationSpotlightSeen {
  type: Actions.NOTIFICATION_SPOTLIGHT_SEEN
}

interface ToggleHideBalances {
  type: Actions.TOGGLE_HIDE_BALANCES
}

interface OptMultichainBeta {
  type: Actions.OPT_MULTICHAIN_BETA
  optedIn: boolean
}

export type ActionTypes =
  | SetAppState
  | SetLoggedIn
  | SetSupportedBiometryType
  | OpenDeepLink
  | SetAnalyticsEnabled
  | SetRequirePinOnAppOpen
  | Lock
  | Unlock
  | SetSessionId
  | OpenUrlAction
  | UpdateConfigValuesAction
  | ActiveScreenChangedAction
  | AppMounted
  | AppUnmounted
  | AndroidMobileServicesAvailabilityChecked
  | PhoneNumberVerificationCompleted
  | PhoneNumberRevoked
  | InviteLinkConsumed
  | HapticFeedbackSet
  | PushNotificationsPermissionChanged
  | inAppReviewRequested
  | NotificationSpotlightSeen
  | ToggleHideBalances
  | OptMultichainBeta
  | DeepLinkDeferred

export const setAppState = (state: string): SetAppState => ({
  type: Actions.SET_APP_STATE,
  state,
})

export const setLoggedIn = (loggedIn: boolean) => ({
  type: Actions.SET_LOGGED_IN,
  loggedIn,
})

export const setSupportedBiometryType = (supportedBiometryType: BIOMETRY_TYPE | null) => ({
  type: Actions.SET_SUPPORTED_BIOMETRY_TYPE,
  supportedBiometryType,
})

export const openDeepLink = (deepLink: string, isSecureOrigin: boolean = false): OpenDeepLink => {
  return {
    type: Actions.OPEN_DEEP_LINK,
    deepLink,
    isSecureOrigin,
  }
}

export const deepLinkDeferred = (deepLink: string, isSecureOrigin: boolean): DeepLinkDeferred => {
  return {
    type: Actions.DEEP_LINK_DEFERRED,
    deepLink,
    isSecureOrigin,
  }
}

export const setAnalyticsEnabled = (enabled: boolean): SetAnalyticsEnabled => ({
  type: Actions.SET_ANALYTICS_ENABLED,
  enabled,
})

export const setRequirePinOnAppOpen = (enabled: boolean): SetRequirePinOnAppOpen => ({
  type: Actions.SET_LOCK_WITH_PIN_ENABLED,
  enabled,
})

export const appLock = (): Lock => ({
  type: Actions.LOCK,
})

export const appUnlock = (): Unlock => ({
  type: Actions.UNLOCK,
})

export const appMounted = (): AppMounted => ({
  type: Actions.APP_MOUNTED,
})

export const appUnmounted = (): AppUnmounted => ({
  type: Actions.APP_UNMOUNTED,
})

export const setSessionId = (sessionId: string) => ({
  type: Actions.SET_SESSION_ID,
  sessionId,
})

export const openUrl = (
  url: string,
  openExternal = false,
  isSecureOrigin = false
): OpenUrlAction => ({
  type: Actions.OPEN_URL,
  url,
  openExternal,
  isSecureOrigin,
})

export const updateRemoteConfigValues = (
  configValues: RemoteConfigValues
): UpdateConfigValuesAction => ({
  type: Actions.UPDATE_REMOTE_CONFIG_VALUES,
  configValues,
})

export const activeScreenChanged = (activeScreen: Screens): ActiveScreenChangedAction => ({
  type: Actions.ACTIVE_SCREEN_CHANGED,
  activeScreen,
})

export const androidMobileServicesAvailabilityChecked = (
  googleIsAvailable: boolean | undefined,
  huaweiIsAvailable: boolean | undefined
): AndroidMobileServicesAvailabilityChecked => ({
  type: Actions.ANDROID_MOBILE_SERVICES_AVAILABILITY_CHECKED,
  googleIsAvailable,
  huaweiIsAvailable,
})

export const phoneNumberVerificationCompleted = (
  e164PhoneNumber: string,
  countryCode: string | null
): PhoneNumberVerificationCompleted => {
  return {
    type: Actions.PHONE_NUMBER_VERIFICATION_COMPLETED,
    e164PhoneNumber,
    countryCode,
  }
}

export const phoneNumberRevoked = (e164PhoneNumber: string): PhoneNumberRevoked => {
  return {
    type: Actions.PHONE_NUMBER_REVOKED,
    e164PhoneNumber,
  }
}

export const inviteLinkConsumed = (inviterAddress: string): InviteLinkConsumed => {
  return {
    type: Actions.INVITE_LINK_CONSUMED,
    inviterAddress,
  }
}

export const hapticFeedbackSet = (hapticFeedbackEnabled: boolean): HapticFeedbackSet => {
  return {
    type: Actions.HAPTIC_FEEDBACK_SET,
    hapticFeedbackEnabled,
  }
}

export const pushNotificationsPermissionChanged = (
  enabled: boolean,
  requestedInApp: boolean
): PushNotificationsPermissionChanged => {
  return {
    type: Actions.PUSH_NOTIFICATIONS_PERMISSION_CHANGED,
    enabled,
    requestedInApp,
  }
}

export const inAppReviewRequested = (
  inAppReviewLastInteractionTimestamp: number
): inAppReviewRequested => {
  return {
    type: Actions.IN_APP_REVIEW_REQUESTED,
    inAppReviewLastInteractionTimestamp,
  }
}

export const notificationSpotlightSeen = (): NotificationSpotlightSeen => {
  return {
    type: Actions.NOTIFICATION_SPOTLIGHT_SEEN,
  }
}

export const toggleHideBalances = (): ToggleHideBalances => {
  return {
    type: Actions.TOGGLE_HIDE_BALANCES,
  }
}

export const optMultichainBeta = (optedIn: boolean): OptMultichainBeta => {
  return {
    type: Actions.OPT_MULTICHAIN_BETA,
    optedIn,
  }
}
