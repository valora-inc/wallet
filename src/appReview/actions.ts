export enum Actions {
  INITIALIZE_APP_REVIEW = 'APP_REVIEW/INITIALIZE_APP_REVIEW',
  UPDATE_APP_REVIEW = 'APP_REVIEW/UPDATE_APP_REVIEW',
}

export interface InitializeAppReviewAction {
  type: Actions.INITIALIZE_APP_REVIEW
}

export interface UpdateAppReviewAction {
  type: Actions.UPDATE_APP_REVIEW
  rated: boolean
}

export type ActionTypes = InitializeAppReviewAction | UpdateAppReviewAction

export function initializeAppReview(): InitializeAppReviewAction {
  return {
    type: Actions.INITIALIZE_APP_REVIEW,
  }
}

export function updateAppReview(rated = false): UpdateAppReviewAction {
  return {
    type: Actions.UPDATE_APP_REVIEW,
    rated,
  }
}
