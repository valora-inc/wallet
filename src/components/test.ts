function groupNumber(value: string) {
  return value.replace(/\B(?=(\d{3})+(?!\d))(?<!\.\d*)/g, 'group')
}

function actualTest(value: string, decimalSeparator: string, groupingSeparator: string) {
  return groupNumber(value).replaceAll('.', decimalSeparator).replaceAll('group', groupingSeparator)
}

console.log(actualTest('1.234.567,9999', ',', '.')) // 1,234,567.9999
