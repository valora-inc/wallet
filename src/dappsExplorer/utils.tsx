import { DappV2WithCategoryNames } from 'src/dapps/types'

/***
 * Calculates a score based on the search term and dapp
 * - +2 if the name matches the search term in a case sensitive way
 * - +1 if the name matches the search term in a case insensitive way
 * - +1 if the description matches the search term in a case insensitive way
 * - +1 if the category matches the search term in a case insensitive way
 * @param dapp: dapp with categoryNames
 * @param searchTerm: the user input search string
 * @returns {number} the score of the dapp 0 - 5
 */
export const calculateSearchScore = (dapp: DappV2WithCategoryNames, searchTerm: string) => {
  if (searchTerm === '') return 0
  const nameMatchingCaseScore = dapp.name.includes(searchTerm) ? 2 : 0
  const nameScore = dapp.name.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const descriptionScore = dapp.description.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const categoryScore = dapp.categoryNames.some((category) =>
    category.toLowerCase().includes(searchTerm.toLowerCase())
  )
    ? 1
    : 0
  return nameMatchingCaseScore + nameScore + descriptionScore + categoryScore
}
