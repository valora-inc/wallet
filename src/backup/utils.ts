import {
  CELO_DERIVATION_PATH_BASE,
  generateKeys,
  MnemonicLanguages,
} from '@celo/cryptographic-utils'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import * as bip39 from 'react-native-bip39'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { useDispatch, useSelector } from 'src/redux/hooks'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { ETHEREUM_DERIVATION_PATH } from 'src/web3/consts'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

const MNEMONIC_STORAGE_KEY = 'mnemonic'

export async function generateKeysFromMnemonic(mnemonic: string) {
  const wordCount = countMnemonicWords(mnemonic)
  const derivationPath = wordCount === 24 ? CELO_DERIVATION_PATH_BASE : ETHEREUM_DERIVATION_PATH
  return generateKeys(mnemonic, undefined, undefined, undefined, bip39, derivationPath)
}

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Recovery Phrase',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function isValidBackupPhrase(phrase: string) {
  const allowedPhraseLengths = [12, 24]
  const phraseLength = countMnemonicWords(formatBackupPhraseOnEdit(phrase))
  return !!phrase && allowedPhraseLengths.includes(phraseLength)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
