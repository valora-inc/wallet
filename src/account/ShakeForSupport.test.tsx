import { fireEvent, render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import ShakeForSupport from 'src/account/ShakeForSupport'
import { AppState } from 'src/app/actions'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { createMockStore } from 'test/utils'

const createStore = (appState: AppState) =>
  createMockStore({
    app: { appState },
  })

describe('ShakeForSupport', () => {
  // beforeEach(() => jest.useRealTimers())

  it('does not render when App not active', async () => {
    const tree = render(
      <Provider store={createStore(AppState.Background)}>
        <ShakeForSupport />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when App is active', async () => {
    const tree = render(
      <Provider store={createStore(AppState.Active)}>
        <ShakeForSupport />
      </Provider>
    )
    expect(tree.queryByTestId('ContactSupportFromShake')).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })

  it('navigates to support screen after button press', async () => {
    const { getByTestId } = render(
      <Provider store={createStore(AppState.Active)}>
        <ShakeForSupport />
      </Provider>
    )
    fireEvent.press(getByTestId('ContactSupportFromShake'))
    expect(navigate).toHaveBeenCalledWith(Screens.SupportContact)
  })
})
