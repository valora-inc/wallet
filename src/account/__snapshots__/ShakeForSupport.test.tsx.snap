// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShakeForSupport does not render when App not active 1`] = `
<Modal
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  panResponderThreshold={4}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  statusBarTranslucent={false}
  supportedOrientations={
    [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  useNativeDriverForBackdrop={true}
  visible={false}
/>
`;
